crates/re_space_view_tensor/src/space_view_class.rs

```diff
@@ -592,58 +583,81 @@ fn paint_axis_names(
     }
 }
 
-fn selectors_ui(ui: &mut egui::Ui, state: &mut ViewTensorState, tensor: &TensorData) {
-    for selector in &state.slice.dim_mapping.selectors {
-        if !selector.visible {
-            continue;
-        }
-
-        let dim = &tensor.shape()[selector.dim_idx];
-        let size = dim.size;
+fn selectors_ui(
+    ctx: &ViewerContext<'_>,
+    ui: &mut egui::Ui,
+    shape: &[TensorDimension],
+    slice_selection: &TensorSliceSelection,
+    slice_property: &ViewProperty<'_>,
+) {
+    let Some(slider) = &slice_selection.slider else {
+        return;
+    };
 
-        let selector_value = state
-            .slice
-            .selector_values
-            .entry(selector.dim_idx)
-            .or_insert_with(|| size / 2); // start in the middle
+    let mut changed_indices = false;
+    let mut indices = slice_selection.indices.clone().unwrap_or_default();
 
-        if size > 0 {
-            *selector_value = selector_value.at_most(size - 1);
+    for index_slider in slider {
+        let dim = &shape[index_slider.dimension as usize];
+        let size = dim.size;
+        if size <= 1 {
+            continue;
         }
 
-        if size > 1 {
-            ui.horizontal(|ui| {
-                let name = dim
-                    .name
-                    .clone()
-                    .map_or_else(|| selector.dim_idx.to_string(), |name| name.to_string());
+        let Some(selector_index) = indices
+            .iter_mut()
+            .find(|i| i.dimension == index_slider.dimension)
+        else {
+            // There should be an entry already via `load_tensor_slice_selection_and_make_valid`
+            continue;
+        };
+        let selector_value = &mut selector_index.index;
```

Comment:
We could make the code in this file a bit more readable by adding helper functions

```suggestion
        let Some(selector_value) = slice_selection.get_slider_index(index_slider.dimension) else {
            // There should be an entry already via `load_tensor_slice_selection_and_make_valid`
            continue;
        };
```

Reply:
it's not a slider index, it's the specific index chosen for a given dimension, but otherwise yes good call :+1: